daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala

```diff
@@ -0,0 +1,188 @@
+
+    def forceRoll(x: Norm.Roll): Node = {
+      x match {
+        case Norm.Roll1(act) =>
+          val child = pushNode(act.node)
+          NodeRollback(children = ImmArray(List(child)))
+        case Norm.Roll2(h, m, t) =>
+          val nodes = List(h.node) ++ m ++ List(t.node)
+          val children = pushNodes(nodes)
+          NodeRollback(children = ImmArray(children))
+      }
+    }
+
+    def forceNorm(x: Norm): Node = {
+      x match {
+        case Norm.Act(node) => node
+        case roll: Norm.Roll => forceRoll(roll)
+      }
+    }
+
+    def forceNorms(xs: List[Norm]): List[Node] = xs.map(forceNorm)
+
+    // makeRoll: encodes the normalization transformation rules:
+    //   rule #1: R [ ] -> ε
+    //   rule #2: R [ R [ xs… ] , ys… ] -> R [ xs… ] , R [ ys… ]
+    //   rule #3: R [ xs… , R [ ys… ] ] ->  R [ xs… , ys… ]
+
+    def makeRoll(norms: List[Norm]): List[Norm] = {
+      caseNorms(norms) match {
+        case Case.Empty =>
+          // normalization rule #1
+          Nil
+
+        case Case.Single(roll: Norm.Roll) =>
+          // normalization rule #2 or #3
+          List(roll)
+
+        case Case.Single(act: Norm.Act) =>
+          // no rule
+          List(Norm.Roll1(act))
+
+        case Case.Multi(h: Norm.Roll, m, t) =>
+          // normalization rule #2
+          h :: makeRoll(m ++ List(t))
+
+        case Case.Multi(h: Norm.Act, m, t: Norm.Roll) =>
+          // normalization rule #3
+          List(pushIntoRoll(h, forceNorms(m), t))
+
+        case Case.Multi(h: Norm.Act, m, t: Norm.Act) =>
+          // no rule
+          List(Norm.Roll2(h, forceNorms(m), t))
+      }
+    }
+
+    def pushIntoRoll(a1: Norm.Act, xs2: List[Node], t: Norm.Roll): Norm.Roll = {
+      t match {
+        case Norm.Roll1(a3) => Norm.Roll2(a1, xs2, a3)
+        case Norm.Roll2(a3, xs4, a5) => Norm.Roll2(a1, xs2 ++ List(a3.node) ++ xs4, a5)
+      }
+    }
+
+    // Here we traverse the original transaction structure.
+    // During the transformation, an original `Node` is mapped into a List[Norm]
+
+    // The List is necessary because the rules can:
+    // (1) drop nodes; (2) combine nodes (3) lift nodes from a lower level to a higher level.
+
+    // TODO: https://github.com/digital-asset/daml/issues/8020
+    // make this traversal code stack-safe
+
+    txOriginal match {
+      case GenTransaction(nodesOriginal, rootsOriginal) =>
+        def traverseNids(nids: List[Nid]): List[Norm] = nids.flatMap(traverseNid)
+        def traverseNid(nid: Nid): List[Norm] = traverseNode(nodesOriginal(nid))
+
+        def traverseNode(node: Node): List[Norm] = {
+          node match {
+
+            case NodeRollback(children) =>
+              makeRoll(traverseNids(children.toList))
+
+            case ex: NodeExercises[_, _] =>
+              val norms = traverseNids(ex.children.toList)
+              val children = pushNodes(forceNorms(norms))
```

Comment:
From what I can tell this might be where you’re breaking the node id order. Now children of an exercise will have a lower node id than their parent. As mentioned on slack, I think we need to preserve node ids = index in preorder at least for Daml Studio and at that point it’s probably easiest to preserve it everywhere (and it’s also useful for debugging over the ledger API).

Reply:
the node ordering is fixed now, and also tested